docs/proposals/20230407-flexible-managed-k8s-endpoints.md

```diff
@@ -0,0 +1,519 @@
+
+```yaml
+apiVersion: cluster.x-k8s.io/v1beta1
+kind: ClusterEndpoint
+spec:
+  hostname: 'name-1234567890.region.elb.amazonaws.com'
+  port: 1234
+  type: ExternalControlPlaneEndpoint
+```
+
+```yaml
+apiVersion: cluster.x-k8s.io/v1beta1
+kind: ClusterEndpoint
+spec:
+  ip: 10.40.85.102
+  port: 1234
+  type: ExternalControlPlaneEndpoint
+```
+
+This type will be the primary way for both `<Infra>Cluster` or `ControlPlane Provider` to communicate the ClusterAddress to "core" Cluster API controllers.
+
+The `<Infra>Cluster` object which is currently using the `spec.controlPlaneEndpoint` for the same scope will continue to work because "core" Cluster API controllers will continue to recognize when this field is set and take care of generating the `ClusterEndpoint` automatically; however this mechanism should be considered as a temporary machinery to migrate to the new CRD, and it will be removed in future versions of Cluster API. In addition, once the legacy behavior is removed, we will deprecate and eventually remove the `spec.controlPlaneEndpoint` field from the `Cluster` CustomResourceDefinition, and recommend that providers do the same for their `<Infra>Cluster` CustomResourceDefinitions as well.
+
+Notes:
+
+- The `hostname` and `ip` fields are mutually exclusive
+- The `type` field has been introduced to allow to extend the usage of this CRD to address https://github.com/kubernetes-sigs/cluster-api/issues/5295 in a future iterations
+- The current implementation originates from the `Cluster.spec.ControlPlaneEndpoint` field, which defines the info we need for this proposal; but in future iterations we might consider to support more addressed or more ports for each ClusterEndpoint, similarly what is implemented in the core v1 Endpoint type.
+
+#### Infra Providers API changes
+
+This proposal does not introduce any breaking changes for the provider's API.
+
+However, Infra providers MUST deprecate the `spec.controlPlaneEndpoint` field in their `<Infra>Cluster` so they can remove it in a future API version.
+
+#### Core Cluster API Controllers changes
+
+- All the controllers working with ClusterClass objects must take into account that the `infrastructure.Ref` field could be omitted; most notably:
+  - The ClusterClass controller must ignore nil `infrastructure.Ref` fields while adding owner references to all the objects referenced by a ClusterClass.
+  - The Topology controller must skip the generation of the `<Infra>Cluster` objects when the `infrastructure.Ref` field in a ClusterClass is empty.
+
+- All the controllers working with Cluster objects must take into account that the `infrastructureRef` field could be omitted; most notably:
+  - The Cluster controller must use skip reconciling this external reference when the `infrastructureRef` is missing; also, the `status.InfrastructureReady` field must be automatically set to true in this case.
+
+- The Cluster controller must reconcile the new `ClusterEndpoint` CR. Please note that:
```

Comment:
This is an implementation detail but since it's specified here, why not have a separate controller for the new CRD?

Reply:
This is definitely an option, but let's defer this to the implementation.
I just checked and the current implementation of the cluster controller already handles a couple of nested reconcile functions, and thus this one can also (alternative to seprated controller) fit into the same pattern

https://github.com/fabriziopandini/cluster-api/blob/f0b2b03bc7cd35d259a64954bb03c66dc4155700/internal/controllers/cluster/cluster_controller.go#L194-L197

Reply:
+ to deferring, in this case let's not specify which controller in the proposal?

Reply:
That is sensible. Updated, PTAL.